### ISSUE TITLE
Authentication token is not sent if terraform well-known endpoint redirects to other domain

### ISSUE THREAD

**Issue by EliaSaSe** [dependabot/dependabot-core#7067]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package ecosystem

terraform

### dependabot.yml content

```yml
version: 2
registries:
  spacelift:
    type: terraform-registry
    url: https://spacelift.io
    token: ${{secrets.SECRET}}
updates:
  - package-ecosystem: terraform
    directory: /
    registries: '*'
    schedule:
      interval: daily
```

### What you expected to see, versus what you actually saw

Hi,

the implementation of the terraform integration implies, that the following issue may be by intention. But maybe the user feedback can be improved:

We are using Spacelift (https://spacelift.io) as terraform module repository: ``url: https://spacelift.io``. We also provide a token.

If dependabot is running with the provided configuration, the well-known endpoint will be called: https://spacelift.io/.well-known/terraform.json
https://github.com/dependabot/dependabot-core/blob/541b99d5acccf991903504d6f8dfdc768afaae04/terraform/lib/dependabot/terraform/registry_client.rb#L150

```json
{
	"login.v1": {
		"authz": "https://spacelift.io/terraform-login",
		"client": "terraform-cli",
		"grant_types": ["authz_code"],
		"token": "https://app.spacelift.io/terraform/token"
	},
	"modules.v1": "https://app.spacelift.io/registry/modules/v1/",
	"providers.v1": "https://app.spacelift.io/registry/providers/v1/",
	"state.v2": "https://app.spacelift.io/state.v2",
	"tfe.v2": "https://app.spacelift.io/state.v2",
	"tfe.v2.1": "https://app.spacelift.io/state.v2",
	"tfe.v2.2": "https://app.spacelift.io/state.v2"
}
```

This json is pointing to a different domain  ``app.spacelift.io``. This causes dependabot to not sent the configured token, because the token is restricted to the domain ``spacelift.io``. No module versions where returned.

https://github.com/dependabot/dependabot-core/blob/541b99d5acccf991903504d6f8dfdc768afaae04/terraform/lib/dependabot/terraform/registry_client.rb#L142

To fix the issue, we configured ``url: https://app.spacelift.io``, which also hosts a .well-known endpoint.

Is this a security feature to not send token to arbitrary domains? So this is not a bug by itself, but it took some time to identify the reason. I would have helped us, if a message like *no token to <module url endpoint> was sent, because for <domain> is no token configured* would have been displayed.

Best regards
Elia

**Comment by EliaSaSe** [comment]
@bodgit Your dependabot.yml looks good, note, that Dependabot does not do recursive searches.   
--> `directory: "/"` if your TF files are not in the root directory, Dependabot will not find them.

But your Log shows that Dependabot was activated, so I would say it's an invalid token. We found, that spacelift always returns a status code 200, even if the access token is invalid. You can not decide if the access token is wrong, a invalid request url is used or if you don't have any modules.

We also had trouble to find the correct value for the secret. We tried until it worked. But I can't remember which one was the correct. @Noahnc, do you remember?

**Comment by Noahnc** [comment]
Hey @bodgit 

As @EliaSaSe pointed out, the spacelift API always returns a 200 status code, even if the request is unauthorized.
When enabling verbose-logging for dependabot, you will see messages like "no version found for module ...".
This is a good indication, that something is wrong regarding authentication, since unauthorized requests respond with an empty list of versions.

If I recall correctly, we used the secret and not the credentials.token from the downloaded config-file.
It is also inportant, that you create a Log-In Policy for the API-Key in spacelift. Otherwise, the Key does not have permission to Log-in and get your modules. The following example is our Policy for the API-Key:
````
# Allow login for the following teams
teams := input.session.teams
allow  { teams[_] == "API" }

# Allow read to root for all api-keys of team "API"
space_read["root"] {
  teams[_] == "API"
}
````
In order for this policy to work, your modules need to be in the root space and the API-Key has to be part of the "API" Group (group names can be added to API-Keys on creation).

**Comment by bodgit** [comment]
Thank you @EliaSaSe & @Noahnc for the tips, I got it working.

I tried using an API key with Admin permissions and this worked, although I wasn't happy giving Dependabot an API key with all those permissions but if it worked 🤷‍♂️ then I saw your advice about the login policy. I checked my existing login policy and sure enough I was only granting permissions on the space(s) for the existing teams that come from our GitHub organization so a non-admin API key couldn't read anything whereas an admin API key probably bypasses the login policy.

I created a new non-admin API key and this time added it to the read-only team that already exists and has `space_read` permission granted by the login policy and this now works and doesn't need scary admin permissions. I might contact Spacelift and suggest adding some documentation on how to set this up as it can be confusing when the registry API just returns 200 responses for everything.

Thanks again!
